## Практическая работа №4 Линейная алгебра

#### 1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\
0 & 0 & 5\\
3 & -4 & 2\\
1 & 6 & 5\\
0 & 1 & 0
\end{pmatrix}.$$

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

In [2]:
A = np.array([[1,2,0],[0,0,5],[3,-4,2],[1,6,5],[0,1,0]])
A

array([[ 1,  2,  0],
       [ 0,  0,  5],
       [ 3, -4,  2],
       [ 1,  6,  5],
       [ 0,  1,  0]])

In [3]:
U, s, W = np.linalg.svd(A)
V = W.T
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

print('U:', U, sep="\n")
print('D:', D, sep="\n")
print('V:', V, sep="\n")

U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]
D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]
V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [7]:
print(U@D@W)  # Проверим

[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1. -0.]]


#### 2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    б) норму Фробениуса.

In [9]:
print('Евклидова норма: ', np.linalg.norm(A, ord=2))

Евклидова норма:  8.824868854820444


евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел $D$. Максимальное значение полученного отношения будет равно максимальному сингулярному числу $\mu_{max}$, и, принимая во внимание факт сортировки по убыванию сингулярных чисел, получим

$$\left \| A \right \|_{E}=\mu_{1}.$$

$$\mu_{1} = 8.82$$
$$\left \| A \right \|_{E}=8.82$$

_норма Фробениуса_, определяется как

$$\left \| A \right \|_{F}=\sqrt{\sum_{i=1}^{m}\sum_{j=1}^{n}a_{ij}^{2}}.$$
В случае, когда известно сингулярное разложение матрицы, ее норма Фробениуса вычисляется как

$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

In [14]:
np_fro_norm = np.linalg.norm(A, ord='fro')
np_fro_norm # для проверки

11.045361017187261

In [30]:
from functools import reduce

fro_norm_v1 = np.sqrt(reduce(lambda x,y: x + y**2, A.flat))
fro_norm_v2 = np.sqrt(np.sum(s**2))

print('норма Фробениуса(Версия 1): ', fro_norm_v1)
print('норма Фробениуса(Версия 2): ', fro_norm_v2)

assert np_fro_norm == fro_norm_v1
assert round(np_fro_norm, 5) == round(fro_norm_v2, 5)

норма Фробениуса(Версия 1):  11.045361017187261
норма Фробениуса(Версия 2):  11.045361017187263
